### PR TITLE
instantiate2 & conditional ibc forwarder instantiation

### PR DESCRIPTION
**Pull request by bekauz**
Begins work towards #138 and introduces instantiate2. Feels like this is getting big enough to open the pr now.

- use instantiate2 for covenant instantiation flow
- only instantiate an ibc forwarder to facilitate party deposit if party is interchain. otherwise the deposit address is just the holder contract.
- bump dependency versions in gh workflows and root level cargo
- move v1 covenant components to a separate dir outside of workspace
- no longer verifying clock whitelisted addresses to be contracts because of instantiate2
- handling `register_fee` params for outgoing ibc messages from neutron (currently pulling from an open pr branch of neutron-sdk, need to revisit after all things are merged on their end)
- interchain router now distributes fallback and target denoms separately

### CODE REVIEW

File: contracts/two-party-pol-covenant/src/contract.rs [hunk 1]
```diff
@@ -11,34 +13,51 @@ use covenant_astroport_liquid_pooler::msg::{
...
+    let precomputed_address = instantiate2_address(&checksum, creator, salt)?;
+
+    Ok(deps.api.addr_humanize(&precomputed_address)?)
+}
+
+pub fn generate_contract_salt(salt_str: &[u8]) -> cosmwasm_std::Binary {
+    let mut hasher = Sha256::new();
+    hasher.update(salt_str);
+    hasher.finalize().to_vec().into()
+}
```
Comment (Art3miX) [inline contracts/two-party-pol-covenant/src/contract.rs:60]: Those helper function should be in "covenant-package", doesn't make sense to copy them around from contract to contract, will be hell to make changes to that in the future.

File: justfile [hunk 2]
```diff
@@ -5,15 +5,15 @@ test:
...
+	cargo clippy --all-targets -- -D warnings
 
 workspace-optimize:
     #!/bin/bash
     docker run --rm -v "$(pwd)":/code \
         --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/target \
         --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry \
         --platform linux/amd64 \
-        cosmwasm/workspace-optimizer:0.12.13
+        cosmwasm/workspace-optimizer:0.14.0
```
Comment (Art3miX) [inline justfile:16]: Already version 0.15 available :)

File: justfile [hunk 3]
```diff
@@ -51,7 +51,7 @@ local-e2e-rebuild TEST: optimize
     fi
     cp -R artifacts/*.wasm {{TEST}}/tests/interchaintest/wasms
     ls {{TEST}}/tests/interchaintest/wasms
-    cd {{TEST}}/tests/interchaintest/ && go test -timeout 50m -v
+    cd {{TEST}}/tests/interchaintest/ && go clean -testcache && go test -timeout 50m -v
```
Comment (Art3miX) [inline justfile:54]: I would not add the testcache to here, this makes the test much slower, I would manually use this command when you know you need to clean cache.

**Review by Art3miX — APPROVED** [review]

**Comment by bekauz** [comment]
Great feedback. Will address it after merging.

**Pull request merged**